Add drag-and-drop feature

pdpino opened the pull request:
Fixes #14 

* See caveats on the issue
* Any improvement on the code or documentation is welcomed!

pdpino commented:
I was thinking the `isDraggable` prop could be: (1) renamed to `areEventsDraggable` (more accurate name) or (2) removed, and make events draggable if and only if `onEventDrag` is provided.

@hoangnm what do you think?

pdpino commented:
* Another improvement I'm thinking: instead of relegating the dragging functionality to the `Draggable` component, implement it on the `Event` component itself, avoiding [this `if` condition](https://github.com/hoangnm/react-native-week-view/pull/72/files?file-filters%5B%5D=.js#diff-4511212c255110f5d3d2037d61adf982112dc92c81b305d367eba89939b0e260R21). Hence, the `Event` would handle both the dragging and pressing functionality.

* I could also refactor the code to be a class component, instead of using hooks (to avoid restricting the versions).

let me know what you think!

I will mark the PR as draft for now.

pdpino commented:
I finished the latest changes:

* I did not use hooks functions, I used a class component instead (`Event` component)
* The `Event` component handles both dragging and pressing behaviors (I removed the `Draggable` component I had created)

I marked the PR as ready :slightly_smiling_face:

hoangnm commented:
Hi @pdpino , I tried to test on local, but the event couldn't move while I'm dragging.

hoangnm commented:
> I was thinking the `areEventsDraggable` prop could be removed, and make events draggable if and only if `onDragEvent` is provided.
> 
> @hoangnm what do you think?

I agree to use `onDragEvent` only to allow dragging.

hoangnm commented:
From my experience, there's also an issue when using PanResponder inside ScrollView, https://stackoverflow.com/questions/51705343/react-native-using-a-panresponder-inside-of-a-scrollview-broken-on-android. I had to turn off the scroll of parent view while an item is dragging inside.

pdpino commented:
> I agree to use onDragEvent only to allow dragging.

Ok, I will change that.

> I tried to test on local, but the event couldn't move while I'm dragging.
> From my experience, there's also an issue when using PanResponder inside ScrollView, https://stackoverflow.com/questions/51705343/react-native-using-a-panresponder-inside-of-a-scrollview-broken-on-android. I had to turn off the scroll of parent view while an item is dragging inside.

Interesting. @hoangnm are you using iOS or android? And what version of react/react-native are you using? I'm using an emulator with android 9, running the `example/` app, with react-native v0.62.0 and react v16.11.0, and the scrolling works. Although, sometimes when I try to drag an event, I end up dragging the whole component instead (vertically or horizontally); not really sure why, and I cannot reproduce it always.

What could be happening is that either the vertical `ScrollView` or the horizontal `VirtualizedList` steal the dragging touch, and do not allow the `Event` to handle it. I have some ideas to solve it:

1. In the [parent `ScrollView`](https://github.com/pdpino/react-native-week-view/blob/feature/drag-and-drop/src/WeekView/WeekView.js#L371) and/or in the [parent `VirtualizedList`](https://github.com/pdpino/react-native-week-view/blob/feature/drag-and-drop/src/WeekView/WeekView.js#L374) set the prop `onMoveShouldSetResponderCapture={() => false}`, to avoid them from stealing the touch in the capturing phase.
2. In the [`PanResponder`](https://github.com/pdpino/react-native-week-view/blob/feature/drag-and-drop/src/Event/Event.js#L22), use `onPanResponderTerminationRequest: (evt, gestureState) => false`, to avoid the other views stealing the touch once it started.
3. Remove [`hasMovedEnough()`](https://github.com/pdpino/react-native-week-view/blob/feature/drag-and-drop/src/Event/Event.js#L11); maybe this is preventing the `Event` component from handling the touch correctly.

Could you try these solutions? I will try to reproduce the behavior and try them.

pdpino commented:
@hoangnm, I removed the `areEventsDraggable` prop.

I'm still not able to reproduce what you are seeing (event not moving on drag). I'm able to drag and move the events :confused:.

I applied the solutions mentioned in the comment above, could you test this again?

---

I'm still seeing something weird, see this gif (I'm not sure if it can be noted from the gif) (this is tested with the latest code from :
![feature-drag-and-drop-v2-bug-capture](https://user-images.githubusercontent.com/20805451/118379974-a27c7580-b5ac-11eb-9658-1d49bd5f5c76.gif)

* I'm able to drag the event multiple times.
* However, when I try to drag it very very fast, **sometimes**, the whole component gets dragged horizontally instead (not that it only happens **sometimes**).
* This has also happened to me **sometimes** scrolling vertically very fast (it does not appear on the gif)

This is weird, but I guess is natural when you have multiple views with drag gestures conflicting. I do not think there is much we can do about it. Do you see something like this?

pdpino commented:
On a separate note, I found and fixed a small bug in this PR (the latest commit). The left is the wrong behavior, and the right is the expected:

![feature-drag-and-drop-v2-bug-splitted-events](https://user-images.githubusercontent.com/20805451/118380373-3bac8b80-b5af-11eb-82d1-50019a6146c3.gif) ![feature-drag-and-drop-v2-bug-splitted-events-expected](https://user-images.githubusercontent.com/20805451/118380708-d8702880-b5b1-11eb-8396-d0131e58f1c1.gif)

* Occurs when an event spanning more than 1 day is dragged (in the example, the "Event 3" goes from saturday 10pm to sunday 2am, lasting 4 hours in total).
* In this example, the second half (the one from sunday 12am to 2am) is dragged
* Result: the event gets shortened to only 2 hours (gif from the left)
* Expected: the event should still last 4 hours, as the original (gif from the right)

### Why did this happen?

* The `onDragEvent()` callback comes with `newStartDate` and `newEndDate` as parameters
* The `newStartDate` gets calculated as the point where the event is dropped. This is correct.
* The `newEndDate` is calculated as `newStartDate + eventDuration`, where `eventDuration` is calculated with the event's `startDate` and `endDate`.
* The duration is incorrect because the `startDate` and `endDate` are changed when the event is splitted in two days (the result is the shortened duration, in the example is 2 hours).
* Hence, the resulting `newEndDate` is incorrect when the event spans more than one day.

### Solution

When sorting the events by date, I added a key `originalDuration`, which saves the original duration of the event (before it is splitted). I think the solution may be a little hacky, but it does the trick. If you think of a better solution let me know!

pdpino commented:
Hi @hoangnm 

Could you check this PR again? I tested it again using the current example app (react 17.0.1, react-native 0.64.2), in an android emulator (android 9) and in an android device (moto g6 android 9), and the dragging works for me.
To drag an event I press it a bit longer than a regular press, but less time than a long-press.

hoangnm commented on src/Event/Event.js:
```diff
@@ -1,38 +1,150 @@
+            width: this.currentWidth,
+            backgroundColor: event.color,
+          },
+          this.translatedByDrag.getTranslateTransform(),
```
From the doc, it seems we have to map with `transform` property, https://reactnative.dev/docs/animatedvaluexy#gettranslatetransform. Is it still working?

pdpino replied:
You are right. For some reason I can't remember, I did it like this (some time ago), and it works. I'm checking into it :thinking:

hoangnm commented on src/Event/Event.styles.js:
```diff
@@ -1,13 +1,17 @@
   },
+  touchableContainer: {
+    flex: 1,
+    alignSelf: 'stretch',
```
Could I know the reason we use `alignSelf`?

pdpino replied:
**If I remember correctly**, `flex: 1` stretches the `TouchableOpacity` in one direction (e.g. the vertical direction, and `alignSelf: 'stretch'` stretches the `TouchableOpacity` in the other direction (horizontally). That way the whole event-box is pressable.

hoangnm commented:
actually, I tried before and it didn't work, it worked when I set the animated values to the transform property. I'm checking again.

hoangnm commented:
Hi @pdpino , I resolved the conflicts and migrated the logic from the class component Event in this PR to functional Event component that you refactored, https://github.com/hoangnm/react-native-week-view/pull/143/files.

hoangnm commented:
The drag-drop with react hook was merged so I'm closing this PR, thanks again @pdpino !
